File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/calendar/IslamicDate.java
package com.byagowi.persiancalendar.calendar;

import com.byagowi.persiancalendar.calendar.islamic.FallbackIslamicConverter;
import com.byagowi.persiancalendar.calendar.islamic.IranianIslamicDateConverter;
import com.byagowi.persiancalendar.calendar.islamic.UmmAlQuraConverter;

/**
 * @author Amir
 * @author ebraminio
 */

public class IslamicDate extends AbstractDate {

    // Converters
    public static boolean useUmmAlQura = false;
    public static int islamicOffset = 0;

    public IslamicDate(int year, int month, int dayOfMonth) {
        super(year, month, dayOfMonth);
    }

    public IslamicDate(long jdn) {
        super(jdn);
    }

    public IslamicDate(AbstractDate date) {
        super(date);
    }

    @Override
    public long toJdn() {
        int year = getYear(), month = getMonth(), day = getDayOfMonth();

        long tableResult = useUmmAlQura
                ? UmmAlQuraConverter.toJdn(year, month, day)
                : IranianIslamicDateConverter.toJdn(year, month, day);

        if (tableResult != -1)
            return tableResult - islamicOffset;

        return FallbackIslamicConverter.toJdn(year, month, day) - islamicOffset;
    }

    @Override
    protected int[] fromJdn(long jdn) {
        jdn += islamicOffset;
        int[] result = useUmmAlQura
                ? UmmAlQuraConverter.fromJdn(jdn)
                : IranianIslamicDateConverter.fromJdn(jdn);

        if (result == null) result = FallbackIslamicConverter.fromJdn(jdn);

        return result;
    }
}

File: shared/src/commonMain/kotlin/io/github/persiancalendar/calendar/IslamicDate.kt
package io.github.persiancalendar.calendar

import io.github.persiancalendar.calendar.util.TwelveMonthsYear.monthStartOfMonthsDistance
import io.github.persiancalendar.calendar.util.TwelveMonthsYear.monthsDistanceTo
import io.github.persiancalendar.calendar.islamic.FallbackIslamicConverter
import io.github.persiancalendar.calendar.islamic.IranianIslamicDateConverter
import io.github.persiancalendar.calendar.islamic.UmmAlQuraConverter

/**
 * @author Amir
 */
class IslamicDate : AbstractDate, YearMonthDate<IslamicDate> {
    constructor(year: Int, month: Int, dayOfMonth: Int) : super(year, month, dayOfMonth)
    constructor(date: AbstractDate) : super(date)
    constructor(jdn: Long) : super(jdn)

    override fun toJdn(): Long {
        val year = year
        val month = month
        val day = dayOfMonth

        val tableResult: Long =
            if (useUmmAlQura) UmmAlQuraConverter.toJdn(year, month, day).toLong()
            else IranianIslamicDateConverter.toJdn(year, month, day)

        return if (tableResult != -1L) tableResult - islamicOffset
        else FallbackIslamicConverter.toJdn(year, month, day) - islamicOffset
    }

    override fun fromJdn(jdn: Long): DateTriplet {
        var jdn = jdn
        jdn += islamicOffset.toLong()
        return (if (useUmmAlQura) {
            UmmAlQuraConverter.fromJdn(jdn)
        } else {
            IranianIslamicDateConverter.fromJdn(jdn)
        }) ?: FallbackIslamicConverter.fromJdn(jdn)
    }

    override fun monthStartOfMonthsDistance(monthsDistance: Int): IslamicDate =
        monthStartOfMonthsDistance(this, monthsDistance, ::IslamicDate)

    override fun monthsDistanceTo(date: IslamicDate): Int = monthsDistanceTo(this, date)

    override fun toString(): String = "IslamicDate($year, $month, $dayOfMonth)"

    companion object {
        // Converters
        var useUmmAlQura = false
        var islamicOffset = 0
    }
}
